### PR TITLE
Replace static IP with hostname for Besu

### PR DESCRIPTION
**Pull request by rranjan3**
- Replace static ip for Besu with respective hostnames
- Cleanup statup/cleanup scripts to report false positives
- Update config files for hostname usage

Signed-off-by: Rajeev Ranjan <rajeev2.ranjan@intel.com>

### CODE REVIEW

File: docs/dev-environments/ethereum/besu/startup.sh [hunk 1]
```diff
@@ -32,10 +33,21 @@ echo ""
 echo "========================================================="
 echo "STEP 3 :: Initialize truffle project and deploy contracts"
 echo "========================================================="
-docker-compose -f docker-compose-truffle.yaml up \
-|| error_exit "Failed to initialize truffle or deploy contracts"
+
+if [ -z "$TCF_HOME" ] || [[ `pwd` != "$TCF_HOME"* ]];
```
Comment (pankajgoyal2) [inline docs/dev-environments/ethereum/besu/startup.sh]: kindly put [[ ]] around $TCF_HOME test condition!

Comment (pankajgoyal2) [reply docs/dev-environments/ethereum/besu/startup.sh]: I think you can modify this condition like this:
-> test for if $TCF_HOME is set or not
       => if not, then throw error and exit
       => if yes, then change directory to $TCF_HOME
I don't think [[ `pwd` != "$TCF_HOME"* ]] check is needed.

Comment (rranjan3) [reply docs/dev-environments/ethereum/besu/startup.sh]: > kindly put [[ ]] around $TCF_HOME test condition!

Thanks. Updated.

Comment (rranjan3) [reply docs/dev-environments/ethereum/besu/startup.sh]: > I think you can modify this condition like this:
> -> test for if $TCF_HOME is set or not
> => if not, then throw error and exit
> => if yes, then change directory to $TCF_HOME
> I don't think [[ `pwd` != "$TCF_HOME"* ]] check is needed.

Clarified  the purpose here. need to run from the same directory but ensuring that it is the right directory and not outside `TCF_HOME`.

Comment (pankajgoyal2) [reply docs/dev-environments/ethereum/besu/startup.sh]: these are not the steps to be done manually but will happen automatically through this script. So, I am suggesting, if TCF_HOME is set, then change directory to TCF_HOME in your script. then you don't need the check if pwd != "$TCF_HOME"

Comment (pankajgoyal2) [reply docs/dev-environments/ethereum/besu/startup.sh]: @rranjan3 Thanks for the clarification. Its clear now. So, I approve the PR.

File: docs/BesuProxyModel.md [hunk 2]
```diff
@@ -1,8 +1,7 @@
 
 **Testing Avalon Proxy model with Hyperledger Besu**
 1.  Set the environment variable ``TCF_HOME`` to the Avalon root directory. Update ``no_proxy`` environment variable if you are behind a proxy
-    server. Add these private IP - ``172.13.0.2, 172.13.0.3, 172.13.0.4, 172.13.0.5``. These are the IP addresses used in the default Besu 
-    network defined in the corresponding [docker-compose file](dev-environments/besu/docker-compose.yaml)
+    server. Add these hostnames - ``node1.avalon.org, rpc.node1.avalon.org, node2.avalon.org, rpc.node2.avalon.org``. These are the IP addresses used in the default Besu network defined in the corresponding [docker-compose file](dev-environments/besu/docker-compose.yaml)
```
Comment (danintel) [inline docs/BesuProxyModel.md]: Your instructions do not show how to discover the IP addresses for 
node1.avalon.org, rpc.node1.avalon.org, node2.avalon.org, rpc.node2.avalon.org.
Before they were IP addresses on the Docker network, 172.13.0.0.  You need to document the commands needed to find out what the IP addresses are or should be.

Comment (danintel) [reply docs/BesuProxyModel.md]: [‎3/‎26/‎2020 10:22 AM]  
Change *.avalon.org to *.avalon.local.  That makes it clear it is a local network.  Also avalon.org is already in use and that may cause confusion.

Comment (rranjan3) [reply docs/BesuProxyModel.md]: Thanks Dan. Updated.

File: docs/dev-environments/ethereum/besu/cleanup.sh [hunk 3]
```diff
@@ -13,10 +13,11 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+exit_status=0
 function error_exit()
 {
     echo "[Error]: " ${*}
-    exit 1
+    exit_status=1
```
Comment (manojgop) [inline docs/dev-environments/ethereum/besu/cleanup.sh]: This function doesn't exit. Can we rename this function in that case. Else let's the function exit on error

File: docs/dev-environments/ethereum/besu/startup.sh [hunk 4]
```diff
@@ -32,10 +33,23 @@ echo ""
...
+
+# Check if TCF_HOME is defined and current working directory
+# is a subtree of TCF_HOME
+if [[ -z "$TCF_HOME" ]] || [[ `pwd` != "$TCF_HOME"* ]];
+then
+    error_exit "TCF_HOME environment variable not defined appropriately. Please check."
+fi
+
+# Verify if contract deployments are successful else log error
+docker-compose -f docker-compose-truffle.yaml up | grep -e "^" -e "Total deployments:   3"
```
Comment (manojgop) [inline docs/dev-environments/ethereum/besu/startup.sh]: Check if the deployment is successful by checking the command that's being executed as part of docker-compose

File: docs/dev-environments/ethereum/besu/docker-compose.yaml [hunk 5]
```diff
@@ -66,18 +66,18 @@ services:
...
     ports:
       - 23001:8555
     volumes:
       - './ethsigner/node2:/var/lib/ethsigner'
     depends_on:
-      - node2.avalon.org
+      - node2.avalon.local
     networks:
       local_net:
         ipv4_address: '172.13.0.5'
```
Comment (manojgop) [inline docs/dev-environments/ethereum/besu/docker-compose.yaml]: Can we remove static IPs in networks

File: docs/dev-environments/ethereum/ganache/cleanup.sh [hunk 6]
```diff
@@ -13,34 +13,51 @@
...
-docker stop local-ganache || error_exit "Failed to stop Ganache container"
+
+# Check if the containers do not exist at all. If they do not exist, cleanup should not fail.
+docker stop local-ganache | grep -e "^" -e "No such container"
+echo $(( $? ^ 1 )) || error_exit "Failed to stop Ganache container"
 echo "Done"
-docker rm local-ganache || error_exit "Failed to delete Ganache container"
+
+docker rm local-ganache | grep -e "^" -e "No such container"
+echo $(( $? ^ 1 )) || error_exit "Failed to delete Ganache container"
```
Comment (manojgop) [inline docs/dev-environments/ethereum/ganache/cleanup.sh]: docker stop local-ganache || true && docker rm local-ganache || true

File: docs/dev-environments/ethereum/besu/docker-compose.yaml (hunk 5 above)
Comment (rranjan3) [reply docs/dev-environments/ethereum/besu/docker-compose.yaml]: Facing issues when I try to remove static IPs. Truffle is not able to run the migration scripts. It hangs forever as suggested by logs-

truffle-envt-besu-setup |
truffle-envt-besu-setup | 1_initial_migration.js
truffle-envt-besu-setup | ======================
truffle-envt-besu-setup |
truffle-envt-besu-setup |    Deploying 'Migrations'
truffle-envt-besu-setup |    ----------------------
truffle-envt-besu-setup |    > transaction hash:    0x1a673f0f62c4f0690eb444d4922cd257b16f72bfff3e4d0362a50f3f9353ede6
truffle-envt-besu-setup | - Blocks: 0            Seconds: 0

**Pull request merged**